lib-grade.rkt

```diff
@@ -136,9 +136,20 @@
      (length (fold-state-errors state))
      (length (fold-state-failures state))))
 
-  (let* ([test-results (call/timeout (λ () (fold-test-results add-result init-state test-suite
-                                                              #:fdown push-suite-name
-                                                              #:fup pop-suite-name)))]
+  ;; run-test-case/timeout : (U String #f) [-> Any] Event -> TestResult
+  ;; runs the test case, but produces (test-error) if test case does not finish before timeout-evt is ready
+  ;; intended to be used with alarm-evt, i.e.
+  ;; (run-test-case/timeout name action (alarm-evt (+ (current-inexact-milliseconds) 2500))
+  (define (run-test-case/timeout name action)
+    (define eng (engine (lambda (b) (run-test-case name action))))
+    (if (engine-run (* 1000 (timeout-in-seconds)) eng)
```

Comment:
Do we want to even construct the engine if `(timeout-in-seconds)` is `#f` ? I was thinking something like the following. Also so that we aren't, in case of a `#f` parameter value, multiplying by a boolean. 

```
    (if (timeout-in-seconds)
        (let ((eng (engine (lambda (p) (run-test-case name action)))))
          (if (engine-run (* 1000 (timeout-in-seconds)) eng)
              (engine-result eng)
              (test-error name (exn:fail "Execution timed out" (current-continuation-marks)))))
        (run-test-case name action))
``` 

A second question. Do timeouts deserve their own classification of things to report, on the same level as successes, failures, errors. Right now in the report we lose the timeout exception you raise, so it's not as informative as it could be.

Reply:
Yeah, this is definitely not optimal. A different report for time outs would be better as well 🤔